{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Preamble\n",
    "\n",
    "Use the environment `shrec`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams.update(plt.rcParamsDefault)\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Benchmark against the ARACNE Algorithm\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.image.AxesImage at 0x164d0b790>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZgAAAGdCAYAAAAv9mXmAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAT8ElEQVR4nO3dfYxV9bno8WcYZANmGBUzRC4jDklPQNAjMtxeAa09Gk4UjCSNrQat0Tap6YggiRGqbaMtTuyLMdGKGdIYWgLyR2ukJ76U2AhSJOKIL2kbSOuJzJESamNmQO8ZHdj3j3s650xH6WzgmbX3+Pkk6w9W1mI9WSR885u1Z6+6crlcDgA4xUYVPQAAI5PAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQIrRw33BY8eOxYEDB6KhoSHq6uqG+/IAnIRyuRyHDx+OyZMnx6hRx1+jDHtgDhw4EM3NzcN9WQBOoa6urpgyZcpxjxn2wDQ0NERExIK4OkbHacN9eQBOQl98HDvimf7/y49n2APztx+LjY7TYnSdwADUlP/69sqhPOLwkB+AFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEgxQkF5rHHHouWlpYYO3ZszJkzJ1566aVTPRcANa7iwGzevDlWrFgR99xzT+zZsycuvfTSuOqqq2L//v0Z8wFQoyoOzEMPPRRf+9rX4utf/3rMmDEjHn744Whubo61a9dmzAdAjaooMB999FF0dnbGwoULB+xfuHBh7Ny58xPP6e3tjZ6engEbACNfRYF577334ujRozFp0qQB+ydNmhQHDx78xHPa29ujsbGxf/M2S4DPhhN6yP/3L5opl8uf+vKZ1atXR3d3d//W1dV1IpcEoMZU9EbLs88+O+rr6wetVg4dOjRoVfM3pVIpSqXSiU8IQE2qaAUzZsyYmDNnTmzdunXA/q1bt8a8efNO6WAA1LaKVjAREStXroybbropWltb45JLLomOjo7Yv39/3HbbbRnzAVCjKg7MV77ylfjrX/8a999/f/z5z3+OWbNmxTPPPBNTp07NmA+AGlVXLpfLw3nBnp6eaGxsjMvj2hhdd9pwXhqAk9RX/jhejKeju7s7JkyYcNxjfRcZACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQIqKv+ySYfTClKInGOyK/yh6gkHqfvO/ih5hkPK/vFv0CFA4KxgAUggMACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQIrRRQ/AcVzxH0VPUBPK//Ju0SMwgvzf51uKHmGQcf/670WPcEKsYABIITAApBAYAFIIDAApBAaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0CKigLT3t4ec+fOjYaGhmhqaoolS5bE3r17s2YDoIZVFJht27ZFW1tb7Nq1K7Zu3Rp9fX2xcOHC+OCDD7LmA6BGVfTCseeee27An5944oloamqKzs7OuOyyy07pYADUtpN6o2V3d3dERJx11lmfekxvb2/09vb2/7mnp+dkLglAjTjhh/zlcjlWrlwZCxYsiFmzZn3qce3t7dHY2Ni/NTc3n+glAaghJxyY22+/Pd58883YtGnTcY9bvXp1dHd3929dXV0nekkAasgJ/Yhs2bJlsWXLlti+fXtMmTLluMeWSqUolUonNBwAtauiwJTL5Vi2bFk89dRT8eKLL0ZLS0vWXADUuIoC09bWFhs3boynn346Ghoa4uDBgxER0djYGOPGjUsZEIDaVNEzmLVr10Z3d3dcfvnlcc455/RvmzdvzpoPgBpV8Y/IAGAofBcZACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQIqTemUywEgz7l//vegRRgwrGABSCAwAKQQGgBQCA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUAgNAitFFDwB8ds1746OiRxhk5z+PKXqEwV6YUvQE/+2D3ohrhnaoFQwAKQQGgBQCA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUAgNACoEBIIXAAJBCYABIcVKBaW9vj7q6ulixYsUpGgeAkeKEA7N79+7o6OiICy+88FTOA8AIcUKBOXLkSCxdujTWrVsXZ5555qmeCYAR4IQC09bWFosWLYorr7zyHx7b29sbPT09AzYARr6KX5n85JNPxmuvvRa7d+8e0vHt7e1x3333VTwYALWtohVMV1dXLF++PDZs2BBjx44d0jmrV6+O7u7u/q2rq+uEBgWgtlS0guns7IxDhw7FnDlz+vcdPXo0tm/fHo8++mj09vZGfX39gHNKpVKUSqVTMy0ANaOiwFxxxRXx1ltvDdh3yy23xPTp0+Puu+8eFBcAPrsqCkxDQ0PMmjVrwL7TTz89Jk6cOGg/AJ9tfpMfgBQVf4rs77344ounYAwARhorGABSCAwAKQQGgBQCA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUJ/1dZAAnauc/jyl6hJrw/Ix/K3qEfj2Hj8WZQzzWCgaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEghcAAkGJ00QMAVJMjz00reoRBHnn//aJH6PefR/oi4u0hHWsFA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFJUHJh33303brzxxpg4cWKMHz8+Lrrooujs7MyYDYAaVtH7YN5///2YP39+fPGLX4xnn302mpqa4k9/+lOcccYZSeMBUKsqCsyDDz4Yzc3N8cQTT/TvO++88071TACMABX9iGzLli3R2toa1113XTQ1NcXs2bNj3bp1xz2nt7c3enp6BmwAjHwVBebtt9+OtWvXxuc+97l4/vnn47bbbos77rgjfvazn33qOe3t7dHY2Ni/NTc3n/TQAFS/unK5XB7qwWPGjInW1tbYuXNn/7477rgjdu/eHS+//PInntPb2xu9vb39f+7p6Ynm5ua4PK6N0XWnncToAKfekeemFT3CINc3V88Hqf7zSF+s+t/boru7OyZMmHDcYytawZxzzjlx/vnnD9g3Y8aM2L9//6eeUyqVYsKECQM2AEa+igIzf/782Lt374B9+/bti6lTp57SoQCofRUF5s4774xdu3bFAw88EH/84x9j48aN0dHREW1tbVnzAVCjKgrM3Llz46mnnopNmzbFrFmz4nvf+148/PDDsXTp0qz5AKhRFf0eTETE4sWLY/HixRmzADCC+C4yAFIIDAApBAaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQVfxcZwEj2T2f8pegRBvm3mWcWPUK/vvLHQz7WCgaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEghcAAkGJ00QMAVJMD/+dw0SOMGFYwAKQQGABSCAwAKQQGgBQCA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUAgNACoEBIEVFgenr64t77703WlpaYty4cTFt2rS4//7749ixY1nzAVCjKnofzIMPPhiPP/54rF+/PmbOnBmvvvpq3HLLLdHY2BjLly/PmhGAGlRRYF5++eW49tprY9GiRRERcd5558WmTZvi1VdfTRkOgNpV0Y/IFixYEC+88ELs27cvIiLeeOON2LFjR1x99dWfek5vb2/09PQM2AAY+Spawdx9993R3d0d06dPj/r6+jh69GisWbMmbrjhhk89p729Pe67776THhSA2lLRCmbz5s2xYcOG2LhxY7z22muxfv36+NGPfhTr16//1HNWr14d3d3d/VtXV9dJDw1A9atoBXPXXXfFqlWr4vrrr4+IiAsuuCDeeeedaG9vj5tvvvkTzymVSlEqlU5+UgBqSkUrmA8//DBGjRp4Sn19vY8pAzBIRSuYa665JtasWRPnnntuzJw5M/bs2RMPPfRQ3HrrrVnzAVCjKgrMI488Et/+9rfjm9/8Zhw6dCgmT54c3/jGN+I73/lO1nwA1Ki6crlcHs4L9vT0RGNjY1we18boutOG89IAnKS+8sfxYjwd3d3dMWHChOMe67vIAEghMACkEBgAUggMACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEgxejhvmC5XI6IiL74OKI83FcH4GT0xccR8d//lx/PsAfm8OHDERGxI54Z7ksDcIocPnw4Ghsbj3tMXXkoGTqFjh07FgcOHIiGhoaoq6s74b+np6cnmpubo6urKyZMmHAKJxxZ3KehcZ+Gxn0ampF8n8rlchw+fDgmT54co0Yd/ynLsK9gRo0aFVOmTDllf9+ECRNG3D9gBvdpaNynoXGfhmak3qd/tHL5Gw/5AUghMACkqNnAlEql+O53vxulUqnoUaqa+zQ07tPQuE9D4z79f8P+kB+Az4aaXcEAUN0EBoAUAgNACoEBIEXNBuaxxx6LlpaWGDt2bMyZMydeeumlokeqKu3t7TF37txoaGiIpqamWLJkSezdu7fosapae3t71NXVxYoVK4oepeq8++67ceONN8bEiRNj/PjxcdFFF0VnZ2fRY1WVvr6+uPfee6OlpSXGjRsX06ZNi/vvvz+OHTtW9GiFqcnAbN68OVasWBH33HNP7NmzJy699NK46qqrYv/+/UWPVjW2bdsWbW1tsWvXrti6dWv09fXFwoUL44MPPih6tKq0e/fu6OjoiAsvvLDoUarO+++/H/Pnz4/TTjstnn322fj9738fP/7xj+OMM84oerSq8uCDD8bjjz8ejz76aPzhD3+IH/zgB/HDH/4wHnnkkaJHK0xNfkz585//fFx88cWxdu3a/n0zZsyIJUuWRHt7e4GTVa+//OUv0dTUFNu2bYvLLrus6HGqypEjR+Liiy+Oxx57LL7//e/HRRddFA8//HDRY1WNVatWxW9/+1s/JfgHFi9eHJMmTYqf/vSn/fu+9KUvxfjx4+PnP/95gZMVp+ZWMB999FF0dnbGwoULB+xfuHBh7Ny5s6Cpql93d3dERJx11lkFT1J92traYtGiRXHllVcWPUpV2rJlS7S2tsZ1110XTU1NMXv27Fi3bl3RY1WdBQsWxAsvvBD79u2LiIg33ngjduzYEVdffXXBkxVn2L/s8mS99957cfTo0Zg0adKA/ZMmTYqDBw8WNFV1K5fLsXLlyliwYEHMmjWr6HGqypNPPhmvvfZa7N69u+hRqtbbb78da9eujZUrV8a3vvWteOWVV+KOO+6IUqkUX/3qV4ser2rcfffd0d3dHdOnT4/6+vo4evRorFmzJm644YaiRytMzQXmb/7+q/7L5fJJff3/SHb77bfHm2++GTt27Ch6lKrS1dUVy5cvj1//+tcxduzYosepWseOHYvW1tZ44IEHIiJi9uzZ8bvf/S7Wrl0rMP/D5s2bY8OGDbFx48aYOXNmvP7667FixYqYPHly3HzzzUWPV4iaC8zZZ58d9fX1g1Yrhw4dGrSqIWLZsmWxZcuW2L59+yl9TcJI0NnZGYcOHYo5c+b07zt69Ghs3749Hn300ejt7Y36+voCJ6wO55xzTpx//vkD9s2YMSN+8YtfFDRRdbrrrrti1apVcf3110dExAUXXBDvvPNOtLe3f2YDU3PPYMaMGRNz5syJrVu3Dti/devWmDdvXkFTVZ9yuRy33357/PKXv4zf/OY30dLSUvRIVeeKK66It956K15//fX+rbW1NZYuXRqvv/66uPyX+fPnD/qI+759+2Lq1KkFTVSdPvzww0Ev4Kqvr/9Mf0y55lYwERErV66Mm266KVpbW+OSSy6Jjo6O2L9/f9x2221Fj1Y12traYuPGjfH0009HQ0ND/4qvsbExxo0bV/B01aGhoWHQM6nTTz89Jk6c6FnV/3DnnXfGvHnz4oEHHogvf/nL8corr0RHR0d0dHQUPVpVueaaa2LNmjVx7rnnxsyZM2PPnj3x0EMPxa233lr0aMUp16if/OQn5alTp5bHjBlTvvjii8vbtm0reqSqEhGfuD3xxBNFj1bVvvCFL5SXL19e9BhV51e/+lV51qxZ5VKpVJ4+fXq5o6Oj6JGqTk9PT3n58uXlc889tzx27NjytGnTyvfcc0+5t7e36NEKU5O/BwNA9au5ZzAA1AaBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEghcAAkEJgAEjx/wAmPF/P7K9B3wAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import numpy as np\n",
    "from scipy.stats import gaussian_kde\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "class ARACNE:\n",
    "    \"\"\"\n",
    "    Optimized ARACNE algorithm for inferring gene regulatory networks.\n",
    "\n",
    "    The approach here is based on the existing open-source implementation:\n",
    "    https://github.com/rugrag/ARACNE-gene-network-inference\n",
    "    However, it has been modified to use vectorized operations and faster KDE\n",
    "\n",
    "    Parameters:\n",
    "        n_permutations (int): Number of permutations to use for the permutation test.\n",
    "        random_state (int): Random seed.\n",
    "\n",
    "    Attributes:\n",
    "        count_matrix (np.ndarray): Count matrix of shape (n_samples, n_genes).\n",
    "        n_samples (int): Number of samples.\n",
    "        n_genes (int): Number of genes.\n",
    "        mutual_info_matrix (np.ndarray): Mutual information matrix of shape \n",
    "            (n_genes, n_genes).\n",
    "        mutual_info_matrix_alternative (np.ndarray): Null distribution of the \n",
    "            mutual information matrix of shape (n_permutations, n_genes, n_genes).\n",
    "        mutual_info_matrix_filtered (np.ndarray): Filtered mutual information \n",
    "            matrix of shape (n_genes, n_genes).\n",
    "        network (np.ndarray): Inferred network as an upper triangular affinity\n",
    "            matrix of shape (n_genes, n_genes).\n",
    "\n",
    "    References:\n",
    "        [1] Margolin, Adam A., et al. BMC Bioinformatics. Vol. 7. 2006.\n",
    "    \"\"\"\n",
    "    def __init__(self, n_permutations=5, random_state=None):\n",
    "        self.n_permutations = n_permutations\n",
    "        self.random_state = random_state\n",
    "        np.random.seed(self.random_state)\n",
    "\n",
    "    def p_kde(self, x, X, h):\n",
    "        kde = gaussian_kde(X, bw_method=h)\n",
    "        return kde(x)[0]\n",
    "\n",
    "    def kernel_mi(self, X, Y):\n",
    "        \"\"\"\n",
    "        Calculate the mutual information between two variables using kernel density \n",
    "        estimation.\n",
    "        \"\"\"\n",
    "        d = 2\n",
    "        nx = len(X)\n",
    "        hx = (4 / (d + 2)) ** (1 / (d + 4)) * nx ** (-1 / (d + 4))\n",
    "\n",
    "        # Compute KDE for X, Y, and their joint distribution\n",
    "        px = np.array([self.p_kde(X[i], X, hx) for i in range(nx)])\n",
    "        py = np.array([self.p_kde(Y[i], Y, hx) for i in range(nx)])\n",
    "        joint_data = np.vstack((X, Y))\n",
    "        pxy = np.array([self.p_kde(np.array([X[i], Y[i]]), joint_data, hx) for i in range(nx)])\n",
    "\n",
    "        # Calculate mutual information\n",
    "        mi_estimation = np.log(pxy / (px * py))\n",
    "        return np.mean(mi_estimation)\n",
    "    \n",
    "    def kernel_mi(self, X, Y, d=2):\n",
    "        nx = len(X)\n",
    "        hx = (4 / (d + 2)) ** (1 / (d + 4)) * nx ** (-1 / (d + 4))\n",
    "\n",
    "        # Compute KDE for X, Y, and their joint distribution\n",
    "        kde_X = gaussian_kde(X, bw_method=hx)\n",
    "        px = kde_X(X)\n",
    "        kde_Y = gaussian_kde(Y, bw_method=hx)\n",
    "        py = kde_Y(Y)\n",
    "        joint_data = np.array([X, Y])\n",
    "        kde_joint = gaussian_kde(joint_data, bw_method=hx)\n",
    "        pxy = kde_joint(joint_data)\n",
    "\n",
    "        # Calculate mutual information\n",
    "        mi_estimation = np.log(pxy / (px * py))\n",
    "        return np.mean(mi_estimation)\n",
    "\n",
    "    def calculate_mutual_info_matrix(self, X):\n",
    "        \"\"\"\n",
    "        Calculate the mutual information matrix for a given data matrix. This \n",
    "        calculation is symmetric, so only the upper triangular portion of the\n",
    "        matrix is stored.\n",
    "        \"\"\"\n",
    "        d = X.shape[1]\n",
    "        mutual_info_matrix = np.zeros((d, d))\n",
    "        for ix in np.arange(d):\n",
    "            for jx in np.arange(ix + 1, d):\n",
    "                mutual_info_matrix[ix, jx] = self.kernel_mi(X[:, ix], X[:, jx])\n",
    "        return mutual_info_matrix\n",
    "\n",
    "    def permutation_test(self):\n",
    "        \"\"\"\n",
    "        Perform a permutation test to estimate the null distribution of the\n",
    "        mutual information matrix.\n",
    "        \"\"\"\n",
    "        mutual_info_matrix_alternative = np.zeros((self.n_permutations, self.n_genes, self.n_genes))\n",
    "        for n in range(self.n_permutations):\n",
    "            shuffled_X = np.apply_along_axis(np.random.permutation, 1, self.count_matrix)\n",
    "            mutual_info_matrix_alternative[n, ...] = self.calculate_mutual_info_matrix(shuffled_X)\n",
    "        return np.mean(mutual_info_matrix_alternative, axis=0)\n",
    "\n",
    "    def filter_mi(self):\n",
    "        \"\"\"\n",
    "        Filter the mutual information matrix based on the alternative\n",
    "        permutation test.\n",
    "        \"\"\"\n",
    "        mutual_info_matrix_filtered = np.copy(self.mutual_info_matrix)\n",
    "        I_0 = np.amax(self.mutual_info_matrix_alternative)\n",
    "        mutual_info_matrix_filtered[mutual_info_matrix_filtered < I_0] = 0\n",
    "        return mutual_info_matrix_filtered\n",
    "\n",
    "    def data_processing_inequality(self, M, i, j, k):\n",
    "        \"\"\"\n",
    "        The Data Processing Inequality (DPI) can be used to filter out edges that\n",
    "        result from indirect relationships, where X -> Y -> Z induces a non-zero\n",
    "        mutual information between X and Z.\n",
    "\n",
    "        The criterion is that if I[i, k] < min(I[i, j], I[j, k]),\n",
    "        then the edge from i to k is filtered.\n",
    "        \"\"\"\n",
    "        dic = {0: (i, j), 1: (i, k), 2: (j, k)}\n",
    "        dpi_list = [M[i, j], M[i, k], M[j, k]]\n",
    "        idx = np.argmin(dpi_list)\n",
    "        return dic[idx]\n",
    "\n",
    "    def filter_loops(self, M):\n",
    "        \"\"\"\n",
    "        Filter out loops from the mutual information matrix based on the\n",
    "        Data Processing Inequality (DPI).\n",
    "        \"\"\"\n",
    "        set_to_zero = []\n",
    "        for i in range(M.shape[0]):\n",
    "            idx_j = np.where(M[i] != 0)[0]  # Indices where M[i] is non-zero\n",
    "            for j in idx_j:\n",
    "                idx_k = np.where(M[j] != 0)[0]  # Indices where M[j] is non-zero\n",
    "                valid_k = idx_k[M[i, idx_k] != 0]\n",
    "                if valid_k.size > 0:\n",
    "                    set_to_zero.extend(\n",
    "                        [self.data_processing_inequality(M, i, j, k) for k in valid_k]\n",
    "                    )\n",
    "\n",
    "        if set_to_zero:\n",
    "            set_to_zero = np.array(set_to_zero).T\n",
    "            M[tuple(set_to_zero)] = 0\n",
    "        \n",
    "        return M\n",
    "\n",
    "    def fit_transform(self, count_matrix):\n",
    "        \"\"\"\n",
    "        Fit the ARACNE algorithm to the count matrix and return the inferred\n",
    "        network as an upper triangular affinity matrix.\n",
    "        \"\"\"\n",
    "        self.count_matrix = count_matrix\n",
    "        self.n_samples, self.n_genes = count_matrix.shape\n",
    "        \n",
    "        # Standardize the input matrix and calculate the mutual information\n",
    "        scaler = StandardScaler()\n",
    "        count_matrix = scaler.fit_transform(count_matrix)\n",
    "        self.mutual_info_matrix = self.calculate_mutual_info_matrix(count_matrix)\n",
    "        \n",
    "        # Perform the permutation test to estimate the null distribution of\n",
    "        # the mutual information matrix\n",
    "        self.mutual_info_matrix_alternative = self.permutation_test()\n",
    "\n",
    "        # Filter the mutual information matrix based on the alternative\n",
    "        # permutation test\n",
    "        self.mutual_info_matrix_filtered = self.filter_mi()\n",
    "        \n",
    "        # Apply the Data Processing Inequality (DPI) to filter out loops\n",
    "        self.network = self.filter_loops(self.mutual_info_matrix_filtered)\n",
    "        \n",
    "        return self.network\n",
    "    \n",
    "aracne = ARACNE(n_permutations=5, random_state=0)\n",
    "\n",
    "np.random.seed(1)\n",
    "X = np.random.rand(100, 10)\n",
    "X0 = np.copy(X)\n",
    "network = aracne.fit_transform(X)\n",
    "plt.imshow(network)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.image.AxesImage at 0x1648f8f10>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZgAAAGdCAYAAAAv9mXmAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAT8ElEQVR4nO3dfYxV9bno8WcYZANmGBUzRC4jDklPQNAjMtxeAa09Gk4UjCSNrQat0Tap6YggiRGqbaMtTuyLMdGKGdIYWgLyR2ukJ76U2AhSJOKIL2kbSOuJzJESamNmQO8ZHdj3j3s650xH6WzgmbX3+Pkk6w9W1mI9WSR885u1Z6+6crlcDgA4xUYVPQAAI5PAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQIrRw33BY8eOxYEDB6KhoSHq6uqG+/IAnIRyuRyHDx+OyZMnx6hRx1+jDHtgDhw4EM3NzcN9WQBOoa6urpgyZcpxjxn2wDQ0NERExIK4OkbHacN9eQBOQl98HDvimf7/y49n2APztx+LjY7TYnSdwADUlP/69sqhPOLwkB+AFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEgxQkF5rHHHouWlpYYO3ZszJkzJ1566aVTPRcANa7iwGzevDlWrFgR99xzT+zZsycuvfTSuOqqq2L//v0Z8wFQoyoOzEMPPRRf+9rX4utf/3rMmDEjHn744Whubo61a9dmzAdAjaooMB999FF0dnbGwoULB+xfuHBh7Ny58xPP6e3tjZ6engEbACNfRYF577334ujRozFp0qQB+ydNmhQHDx78xHPa29ujsbGxf/M2S4DPhhN6yP/3L5opl8uf+vKZ1atXR3d3d//W1dV1IpcEoMZU9EbLs88+O+rr6wetVg4dOjRoVfM3pVIpSqXSiU8IQE2qaAUzZsyYmDNnTmzdunXA/q1bt8a8efNO6WAA1LaKVjAREStXroybbropWltb45JLLomOjo7Yv39/3HbbbRnzAVCjKg7MV77ylfjrX/8a999/f/z5z3+OWbNmxTPPPBNTp07NmA+AGlVXLpfLw3nBnp6eaGxsjMvj2hhdd9pwXhqAk9RX/jhejKeju7s7JkyYcNxjfRcZACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQIqKv+ySYfTClKInGOyK/yh6gkHqfvO/ih5hkPK/vFv0CFA4KxgAUggMACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQIrRRQ/AcVzxH0VPUBPK//Ju0SMwgvzf51uKHmGQcf/670WPcEKsYABIITAApBAYAFIIDAApBAaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0CKigLT3t4ec+fOjYaGhmhqaoolS5bE3r17s2YDoIZVFJht27ZFW1tb7Nq1K7Zu3Rp9fX2xcOHC+OCDD7LmA6BGVfTCseeee27An5944oloamqKzs7OuOyyy07pYADUtpN6o2V3d3dERJx11lmfekxvb2/09vb2/7mnp+dkLglAjTjhh/zlcjlWrlwZCxYsiFmzZn3qce3t7dHY2Ni/NTc3n+glAaghJxyY22+/Pd58883YtGnTcY9bvXp1dHd3929dXV0nekkAasgJ/Yhs2bJlsWXLlti+fXtMmTLluMeWSqUolUonNBwAtauiwJTL5Vi2bFk89dRT8eKLL0ZLS0vWXADUuIoC09bWFhs3boynn346Ghoa4uDBgxER0djYGOPGjUsZEIDaVNEzmLVr10Z3d3dcfvnlcc455/RvmzdvzpoPgBpV8Y/IAGAofBcZACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQIqTemUywEgz7l//vegRRgwrGABSCAwAKQQGgBQCA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUAgNAitFFDwB8ds1746OiRxhk5z+PKXqEwV6YUvQE/+2D3ohrhnaoFQwAKQQGgBQCA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUAgNACoEBIIXAAJBCYABIcVKBaW9vj7q6ulixYsUpGgeAkeKEA7N79+7o6OiICy+88FTOA8AIcUKBOXLkSCxdujTWrVsXZ5555qmeCYAR4IQC09bWFosWLYorr7zyHx7b29sbPT09AzYARr6KX5n85JNPxmuvvRa7d+8e0vHt7e1x3333VTwYALWtohVMV1dXLF++PDZs2BBjx44d0jmrV6+O7u7u/q2rq+uEBgWgtlS0guns7IxDhw7FnDlz+vcdPXo0tm/fHo8++mj09vZGfX39gHNKpVKUSqVTMy0ANaOiwFxxxRXx1ltvDdh3yy23xPTp0+Puu+8eFBcAPrsqCkxDQ0PMmjVrwL7TTz89Jk6cOGg/AJ9tfpMfgBQVf4rs77344ounYAwARhorGABSCAwAKQQGgBQCA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUJ/1dZAAnauc/jyl6hJrw/Ix/K3qEfj2Hj8WZQzzWCgaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEghcAAkGJ00QMAVJMjz00reoRBHnn//aJH6PefR/oi4u0hHWsFA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFJUHJh33303brzxxpg4cWKMHz8+Lrrooujs7MyYDYAaVtH7YN5///2YP39+fPGLX4xnn302mpqa4k9/+lOcccYZSeMBUKsqCsyDDz4Yzc3N8cQTT/TvO++88071TACMABX9iGzLli3R2toa1113XTQ1NcXs2bNj3bp1xz2nt7c3enp6BmwAjHwVBebtt9+OtWvXxuc+97l4/vnn47bbbos77rgjfvazn33qOe3t7dHY2Ni/NTc3n/TQAFS/unK5XB7qwWPGjInW1tbYuXNn/7477rgjdu/eHS+//PInntPb2xu9vb39f+7p6Ynm5ua4PK6N0XWnncToAKfekeemFT3CINc3V88Hqf7zSF+s+t/boru7OyZMmHDcYytawZxzzjlx/vnnD9g3Y8aM2L9//6eeUyqVYsKECQM2AEa+igIzf/782Lt374B9+/bti6lTp57SoQCofRUF5s4774xdu3bFAw88EH/84x9j48aN0dHREW1tbVnzAVCjKgrM3Llz46mnnopNmzbFrFmz4nvf+148/PDDsXTp0qz5AKhRFf0eTETE4sWLY/HixRmzADCC+C4yAFIIDAApBAaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQVfxcZwEj2T2f8pegRBvm3mWcWPUK/vvLHQz7WCgaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEghcAAkGJ00QMAVJMD/+dw0SOMGFYwAKQQGABSCAwAKQQGgBQCA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUAgNACoEBIEVFgenr64t77703WlpaYty4cTFt2rS4//7749ixY1nzAVCjKnofzIMPPhiPP/54rF+/PmbOnBmvvvpq3HLLLdHY2BjLly/PmhGAGlRRYF5++eW49tprY9GiRRERcd5558WmTZvi1VdfTRkOgNpV0Y/IFixYEC+88ELs27cvIiLeeOON2LFjR1x99dWfek5vb2/09PQM2AAY+Spawdx9993R3d0d06dPj/r6+jh69GisWbMmbrjhhk89p729Pe67776THhSA2lLRCmbz5s2xYcOG2LhxY7z22muxfv36+NGPfhTr16//1HNWr14d3d3d/VtXV9dJDw1A9atoBXPXXXfFqlWr4vrrr4+IiAsuuCDeeeedaG9vj5tvvvkTzymVSlEqlU5+UgBqSkUrmA8//DBGjRp4Sn19vY8pAzBIRSuYa665JtasWRPnnntuzJw5M/bs2RMPPfRQ3HrrrVnzAVCjKgrMI488Et/+9rfjm9/8Zhw6dCgmT54c3/jGN+I73/lO1nwA1Ki6crlcHs4L9vT0RGNjY1we18boutOG89IAnKS+8sfxYjwd3d3dMWHChOMe67vIAEghMACkEBgAUggMACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEghcAAkEJgAEghMACkEBgAUggMACkEBoAUAgNACoEBIIXAAJBCYABIITAApBAYAFIIDAApBAaAFAIDQAqBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEgxejhvmC5XI6IiL74OKI83FcH4GT0xccR8d//lx/PsAfm8OHDERGxI54Z7ksDcIocPnw4Ghsbj3tMXXkoGTqFjh07FgcOHIiGhoaoq6s74b+np6cnmpubo6urKyZMmHAKJxxZ3KehcZ+Gxn0ampF8n8rlchw+fDgmT54co0Yd/ynLsK9gRo0aFVOmTDllf9+ECRNG3D9gBvdpaNynoXGfhmak3qd/tHL5Gw/5AUghMACkqNnAlEql+O53vxulUqnoUaqa+zQ07tPQuE9D4z79f8P+kB+Az4aaXcEAUN0EBoAUAgNACoEBIEXNBuaxxx6LlpaWGDt2bMyZMydeeumlokeqKu3t7TF37txoaGiIpqamWLJkSezdu7fosapae3t71NXVxYoVK4oepeq8++67ceONN8bEiRNj/PjxcdFFF0VnZ2fRY1WVvr6+uPfee6OlpSXGjRsX06ZNi/vvvz+OHTtW9GiFqcnAbN68OVasWBH33HNP7NmzJy699NK46qqrYv/+/UWPVjW2bdsWbW1tsWvXrti6dWv09fXFwoUL44MPPih6tKq0e/fu6OjoiAsvvLDoUarO+++/H/Pnz4/TTjstnn322fj9738fP/7xj+OMM84oerSq8uCDD8bjjz8ejz76aPzhD3+IH/zgB/HDH/4wHnnkkaJHK0xNfkz585//fFx88cWxdu3a/n0zZsyIJUuWRHt7e4GTVa+//OUv0dTUFNu2bYvLLrus6HGqypEjR+Liiy+Oxx57LL7//e/HRRddFA8//HDRY1WNVatWxW9/+1s/JfgHFi9eHJMmTYqf/vSn/fu+9KUvxfjx4+PnP/95gZMVp+ZWMB999FF0dnbGwoULB+xfuHBh7Ny5s6Cpql93d3dERJx11lkFT1J92traYtGiRXHllVcWPUpV2rJlS7S2tsZ1110XTU1NMXv27Fi3bl3RY1WdBQsWxAsvvBD79u2LiIg33ngjduzYEVdffXXBkxVn2L/s8mS99957cfTo0Zg0adKA/ZMmTYqDBw8WNFV1K5fLsXLlyliwYEHMmjWr6HGqypNPPhmvvfZa7N69u+hRqtbbb78da9eujZUrV8a3vvWteOWVV+KOO+6IUqkUX/3qV4ser2rcfffd0d3dHdOnT4/6+vo4evRorFmzJm644YaiRytMzQXmb/7+q/7L5fJJff3/SHb77bfHm2++GTt27Ch6lKrS1dUVy5cvj1//+tcxduzYosepWseOHYvW1tZ44IEHIiJi9uzZ8bvf/S7Wrl0rMP/D5s2bY8OGDbFx48aYOXNmvP7667FixYqYPHly3HzzzUWPV4iaC8zZZ58d9fX1g1Yrhw4dGrSqIWLZsmWxZcuW2L59+yl9TcJI0NnZGYcOHYo5c+b07zt69Ghs3749Hn300ejt7Y36+voCJ6wO55xzTpx//vkD9s2YMSN+8YtfFDRRdbrrrrti1apVcf3110dExAUXXBDvvPNOtLe3f2YDU3PPYMaMGRNz5syJrVu3Dti/devWmDdvXkFTVZ9yuRy33357/PKXv4zf/OY30dLSUvRIVeeKK66It956K15//fX+rbW1NZYuXRqvv/66uPyX+fPnD/qI+759+2Lq1KkFTVSdPvzww0Ev4Kqvr/9Mf0y55lYwERErV66Mm266KVpbW+OSSy6Jjo6O2L9/f9x2221Fj1Y12traYuPGjfH0009HQ0ND/4qvsbExxo0bV/B01aGhoWHQM6nTTz89Jk6c6FnV/3DnnXfGvHnz4oEHHogvf/nL8corr0RHR0d0dHQUPVpVueaaa2LNmjVx7rnnxsyZM2PPnj3x0EMPxa233lr0aMUp16if/OQn5alTp5bHjBlTvvjii8vbtm0reqSqEhGfuD3xxBNFj1bVvvCFL5SXL19e9BhV51e/+lV51qxZ5VKpVJ4+fXq5o6Oj6JGqTk9PT3n58uXlc889tzx27NjytGnTyvfcc0+5t7e36NEKU5O/BwNA9au5ZzAA1AaBASCFwACQQmAASCEwAKQQGABSCAwAKQQGgBQCA0AKgQEghcAAkEJgAEjx/wAmPF/P7K9B3wAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "aracne = ARACNE(n_permutations=5, random_state=0)\n",
    "\n",
    "np.random.seed(1)\n",
    "X = np.random.rand(100, 10)\n",
    "X0 = np.copy(X)\n",
    "network = aracne.fit_transform(X)\n",
    "plt.imshow(network)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "all_datasets = sorted(glob.glob(\"data/combination_regulation/*\"))\n",
    "all_seed_vals = [int(x.split(\"_\")[-1].split(\".\")[0]) for x in all_datasets]\n",
    "\n",
    "for seed_val in all_seed_vals[-100:]:\n",
    "\n",
    "    ## Load the dataset, the true driver, and the indices into the original regulatory network\n",
    "    y_driver = np.load(f\"data/combination_regulation/y_driver_{seed_val}.npy\", allow_pickle=True)\n",
    "    X_measure = np.load(f\"data/combination_regulation/X_measure_{seed_val}.npy\", allow_pickle=True)\n",
    "    true_indices = np.load(f\"data/combination_regulation/simplex_indices_{seed_val}.npy\", allow_pickle=True)[1:]\n",
    "\n",
    "    aracne = ARACNE(n_permutations=5, random_state=0)\n",
    "\n",
    "    network = aracne.fit_transform(X)\n",
    "\n",
    "    top_indices = np.argsort(np.sum(np.abs(network), axis=1))[::-1]\n",
    "\n",
    "    # save indices to text file\n",
    "    np.savetxt(f\"./dump_results/aracne_output/indices_{seed_val}.txt\", top_indices, fmt=\"%d\")\n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([4, 6])"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directed: False\n",
      "Number of nodes: 4441\n",
      "Number of edges: 12864\n",
      "Adjacency matrix stored as SciPy CSR sparse matrix with shape (4441, 4441)\n",
      "Nodes: ['YAL051W', 'YAL016W', 'YAL034WA', 'YAL035CA', 'YAL035W', 'YAL036C', 'YBL089W', 'YBR019C', 'YBR020W', 'YBR127C'] ...\n",
      "Edges: [('YAL051W', 'YAL016W'), ('YAL051W', 'YAL034WA'), ('YAL051W', 'YAL035CA'), ('YAL051W', 'YAL035W'), ('YAL051W', 'YAL036C'), ('YAL051W', 'YBL089W'), ('YAL051W', 'YBR019C'), ('YAL051W', 'YBR020W'), ('YAL051W', 'YBR127C'), ('YAL051W', 'YBR150C')] ...\n",
      "Adjacency Matrix (SciPy Sparse):\n",
      "  (0, 1)\t1\n",
      "  (0, 2)\t1\n",
      "  (0, 3)\t1\n",
      "  (0, 4)\t1\n",
      "  (0, 5)\t1\n",
      "  (0, 6)\t1\n",
      "  (0, 7)\t1\n",
      "  (0, 8)\t1\n",
      "  (0, 9)\t1\n",
      "  (0, 10)\t1\n",
      "  (0, 11)\t1\n",
      "  (0, 12)\t1\n",
      "  (0, 13)\t1\n",
      "  (0, 14)\t1\n",
      "  (0, 15)\t1\n",
      "  (0, 16)\t1\n",
      "  (0, 17)\t1\n",
      "  (0, 18)\t1\n",
      "  (0, 19)\t1\n",
      "  (0, 20)\t1\n",
      "  (0, 21)\t1\n",
      "  (0, 22)\t1\n",
      "  (0, 23)\t1\n",
      "  (0, 24)\t1\n",
      "  (0, 25)\t1\n",
      "  :\t:\n",
      "  (4439, 3287)\t1\n",
      "  (4440, 94)\t1\n",
      "  (4440, 103)\t1\n",
      "  (4440, 106)\t1\n",
      "  (4440, 124)\t1\n",
      "  (4440, 126)\t1\n",
      "  (4440, 169)\t1\n",
      "  (4440, 179)\t1\n",
      "  (4440, 182)\t1\n",
      "  (4440, 419)\t1\n",
      "  (4440, 557)\t1\n",
      "  (4440, 653)\t1\n",
      "  (4440, 693)\t1\n",
      "  (4440, 934)\t1\n",
      "  (4440, 1177)\t1\n",
      "  (4440, 1370)\t1\n",
      "  (4440, 1404)\t1\n",
      "  (4440, 1405)\t1\n",
      "  (4440, 1539)\t1\n",
      "  (4440, 1839)\t1\n",
      "  (4440, 2441)\t1\n",
      "  (4440, 2717)\t1\n",
      "  (4440, 2718)\t1\n",
      "  (4440, 2883)\t1\n",
      "  (4440, 3016)\t1\n",
      "Shortest path from YAL051W to YER015W: ['YAL051W', 'YER015W']\n",
      "Degree Distribution: {61: 1, 4: 350, 3: 624, 1: 1556, 2: 1062, 9: 61, 10: 46, 6: 132, 8: 55, 207: 1, 7: 111, 5: 212, 149: 1, 30: 3, 13: 12, 11: 24, 14: 9, 16: 19, 117: 1, 27: 5, 19: 4, 22: 3, 12: 12, 25: 2, 28: 3, 18: 2, 17: 4, 15: 11, 46: 2, 209: 1, 148: 1, 29: 2, 203: 1, 87: 2, 85: 1, 23: 5, 279: 1, 73: 1, 21: 5, 154: 1, 110: 1, 26: 3, 54: 2, 52: 2, 42: 1, 91: 1, 56: 2, 264: 1, 49: 1, 47: 1, 152: 1, 126: 2, 136: 1, 135: 1, 246: 1, 160: 1, 206: 1, 194: 1, 300: 1, 356: 1, 63: 3, 102: 1, 55: 2, 216: 1, 204: 1, 78: 1, 150: 1, 140: 1, 62: 2, 278: 1, 180: 1, 252: 1, 96: 1, 163: 1, 20: 1, 35: 1, 38: 1, 67: 1, 76: 2, 231: 2, 51: 2, 168: 1, 70: 2, 41: 1, 157: 1, 285: 1, 132: 1, 40: 4, 109: 1, 65: 1, 39: 1, 115: 1, 53: 1, 92: 1, 357: 1, 249: 1, 74: 1, 219: 1, 185: 2, 50: 2, 171: 1, 235: 1, 120: 1, 255: 1, 228: 1, 31: 3, 37: 1, 186: 1, 60: 1, 71: 1, 24: 1}\n",
      "Degrees for node MAL61: 1\n",
      "Degrees for node MEL1: 1\n",
      "Degrees for node STA1: 5\n",
      "Degrees for node STA2: 4\n",
      "Degrees for node STA3: 3\n",
      "Degrees for node YAL001C: 1\n",
      "Degrees for node YAL003W: 3\n",
      "Degrees for node YAL004W: 2\n",
      "Degrees for node YAL005C: 6\n",
      "Degrees for node YAL007C: 1\n"
     ]
    }
   ],
   "source": [
    "import networkx as nx\n",
    "from scipy.sparse import csr_matrix\n",
    "import numpy as np\n",
    "\n",
    "class GeneNetwork:\n",
    "    def __init__(self, filepath, delimiter='\\t', header=True, directed=False):\n",
    "        \"\"\"\n",
    "        Initializes the GeneNetwork by reading the file and creating a graph.\n",
    "        \n",
    "        Parameters:\n",
    "        - filepath (str): Path to the gene interaction file.\n",
    "        - delimiter (str): Delimiter used in the file (default is tab).\n",
    "        - header (bool): Indicates if the file has a header row (default is True).\n",
    "        - directed (bool): Specifies if the graph is directed (default is False).\n",
    "        \"\"\"\n",
    "        self.filepath = filepath\n",
    "        self.delimiter = delimiter\n",
    "        self.header = header\n",
    "        self.directed = directed\n",
    "        self.graph = self._create_graph()\n",
    "        self.adj_matrix = self._create_adjacency_matrix()\n",
    "    \n",
    "    def _create_graph(self):\n",
    "        \"\"\"\n",
    "        Reads the file and creates a NetworkX graph.\n",
    "        \n",
    "        Returns:\n",
    "        - G (networkx.Graph or networkx.DiGraph): The gene interaction graph.\n",
    "        \"\"\"\n",
    "        G = nx.DiGraph() if self.directed else nx.Graph()\n",
    "        \n",
    "        with open(self.filepath, 'r') as file:\n",
    "            for idx, line in enumerate(file):\n",
    "                if self.header and idx == 0:\n",
    "                    continue  # Skip header line\n",
    "                parts = line.strip().split(self.delimiter)\n",
    "                if len(parts) != 2:\n",
    "                    continue  # Skip malformed lines\n",
    "                tf, tg = parts\n",
    "                G.add_edge(tf, tg)\n",
    "        \n",
    "        return G\n",
    "    \n",
    "    def _create_adjacency_matrix(self):\n",
    "        \"\"\"\n",
    "        Creates the adjacency matrix of the graph as a SciPy sparse CSR matrix.\n",
    "        \n",
    "        Returns:\n",
    "        - adj_matrix (scipy.sparse.csr_matrix): The adjacency matrix in CSR format.\n",
    "        \"\"\"\n",
    "        # Using NetworkX's built-in function to create a sparse adjacency matrix\n",
    "        adj_matrix = nx.to_scipy_sparse_array(\n",
    "            self.graph,\n",
    "            format='csr',\n",
    "            dtype=np.int8  # Use a small integer type to save memory\n",
    "        )\n",
    "        return adj_matrix\n",
    "    \n",
    "\n",
    "    def __str__(self):\n",
    "        \"\"\"\n",
    "        Prints basic information about the graph.\n",
    "        \"\"\"\n",
    "        return (f\"Directed: {self.directed}\\n\"\n",
    "                f\"Number of nodes: {self.graph.number_of_edges()}\\n\"\n",
    "                f\"Number of edges: {self.graph.number_of_edges()}\\n\"\n",
    "                f\"Adjacency matrix stored as SciPy CSR sparse matrix with shape {self.adj_matrix.shape}\")\n",
    "    \n",
    "    def get_node_list(self):\n",
    "        \"\"\"\n",
    "        Returns a sorted list of nodes in the graph.\n",
    "        \n",
    "        Returns:\n",
    "        - nodes (list): Sorted list of node identifiers.\n",
    "        \"\"\"\n",
    "        return sorted(self.graph.nodes())\n",
    "    \n",
    "    def get_edges(self):\n",
    "        \"\"\"\n",
    "        Returns a list of edges in the graph.\n",
    "        \n",
    "        Returns:\n",
    "        - edges (list of tuples): List containing edge tuples (u, v).\n",
    "        \"\"\"\n",
    "        return list(self.graph.edges())\n",
    "    \n",
    "    def find_shortest_path(self, source, target):\n",
    "        \"\"\"\n",
    "        Finds the shortest path between two nodes using NetworkX's shortest path algorithms.\n",
    "        \n",
    "        Parameters:\n",
    "        - source (str): Source node.\n",
    "        - target (str): Target node.\n",
    "        \n",
    "        Returns:\n",
    "        - path (list): List of nodes representing the shortest path.\n",
    "        \"\"\"\n",
    "        try:\n",
    "            path = nx.shortest_path(self.graph, source=source, target=target)\n",
    "            return path\n",
    "        except nx.NetworkXNoPath:\n",
    "            print(f\"No path between {source} and {target}.\")\n",
    "            return None\n",
    "        except nx.NodeNotFound as e:\n",
    "            print(e)\n",
    "            return None\n",
    "    \n",
    "    def degree_distribution(self):\n",
    "        \"\"\"\n",
    "        Computes the degree distribution of the graph.\n",
    "        \n",
    "        Returns:\n",
    "        - degree_count (dict): Dictionary mapping degree to number of nodes with that degree.\n",
    "        \"\"\"\n",
    "        degrees = dict(self.graph.degree())\n",
    "        degree_count = {}\n",
    "        for degree in degrees.values():\n",
    "            degree_count[degree] = degree_count.get(degree, 0) + 1\n",
    "        return degree_count\n",
    "    \n",
    "    def get_node_degrees(self):\n",
    "        \"\"\"\n",
    "        Computes the degree of each node in the graph.\n",
    "        \n",
    "        For undirected graphs, returns the degree.\n",
    "        For directed graphs, returns both in-degree and out-degree.\n",
    "        \n",
    "        Returns:\n",
    "        - degrees (dict): \n",
    "            - If undirected: {node: degree}\n",
    "            - If directed: {node: {'in_degree': x, 'out_degree': y}}\n",
    "        \"\"\"\n",
    "        degrees = {}\n",
    "        if self.directed:\n",
    "            in_degrees = dict(self.graph.in_degree())\n",
    "            out_degrees = dict(self.graph.out_degree())\n",
    "            for node in self.graph.nodes():\n",
    "                degrees[node] = {\n",
    "                    'in_degree': in_degrees.get(node, 0),\n",
    "                    'out_degree': out_degrees.get(node, 0)\n",
    "                }\n",
    "        else:\n",
    "            degrees = dict(self.graph.degree())\n",
    "        return degrees\n",
    "\n",
    "# Example Usage\n",
    "if __name__ == \"__main__\":\n",
    "    # Assuming the file is named 'gene_network.txt' and is in the current directory\n",
    "    filepath = '/Users/william/program_repos/dygene/data/yeast_transcriptional_network_Balaji2006.txt'\n",
    "    \n",
    "    # Initialize the GeneNetwork (assuming undirected graph)\n",
    "    gene_network = GeneNetwork(filepath, directed=False)\n",
    "    \n",
    "    # Display basic information\n",
    "    gene_network.display_basic_info()\n",
    "    \n",
    "    # Access the NetworkX graph\n",
    "    G = gene_network.get_networkx_graph()\n",
    "    print(\"Nodes:\", list(G.nodes())[:10], \"...\")  # Display first 10 nodes\n",
    "    print(\"Edges:\", list(G.edges())[:10], \"...\")  # Display first 10 edges\n",
    "    \n",
    "    # Get adjacency matrix as SciPy sparse matrix\n",
    "    adj_sp = gene_network.get_adjacency_matrix_sparse()\n",
    "    print(\"Adjacency Matrix (SciPy Sparse):\")\n",
    "    print(adj_sp)\n",
    "    \n",
    "    # Example: Find shortest path between two genes\n",
    "    source_gene = 'YAL051W'\n",
    "    target_gene = 'YER015W'\n",
    "    path = gene_network.find_shortest_path(source_gene, target_gene)\n",
    "    if path:\n",
    "        print(f\"Shortest path from {source_gene} to {target_gene}: {path}\")\n",
    "    \n",
    "    # Example: Get degree distribution\n",
    "    degree_dist = gene_network.degree_distribution()\n",
    "    print(\"Degree Distribution:\", degree_dist)\n",
    "    \n",
    "    # Example: Get degrees of all nodes\n",
    "    node_degrees = gene_network.get_node_degrees()\n",
    "    # Display degrees of first 10 nodes\n",
    "    first_10_nodes = list(gene_network.get_node_list())[:10]\n",
    "    for node in first_10_nodes:\n",
    "        print(f\"Degrees for node {node}: {node_degrees[node]}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.4 ('shrec2')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "e6225c57107570fd6f8cc9d77f9cc493b42f871d2f9e87a02ce72bc3dd2bd7fd"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
